### PR TITLE
[Polymetis] Add IK to EE movement commands (EE tracking improvement #3)

### PR DESCRIPTION
**Pull request by exhaustin**
# Description

Change EE pose commands (`move_to_ee_pose` and `update_ee_pose_desired`) to use the same adaptive joint PD controller created in #1255.

Before: desired_ee_pose --> operational space PD (bad tracking performance)
After: desired_ee_pose --> IK --> desired joint_pose --> adaptive joint PD

### Changes
- Add IK as a method to `RobotInterface`, in the form of a wrapper around the current IK in `PinocchioRobotModel`.
- Rewrite controllers in `move_to_ee_pose` and `start_cartesian_impedance` to use IK + adaptive joint control.
- Change default time to go within RobotInterface as the adaptive time to go value is now accurately estimated given all movements are in the joint space.

### Notes regarding Trac-IK
Originally attempted to use [tracikpy](https://github.com/mjd3/tracikpy), but some dependencies (`libkdl-parser-dev`, `libnlopt-dev`, and `liborocos-kdl-dev`) are only available in Debian packages. 

It is still beneficial to replace the current Pinocchio IK with Trac-IK for: 
1. Slightly better performance
2. Solutions that take into account joint limits
3. The ability to optimize for less singularity. 

It might be possible to build from source (using catkin) into a conda package, but this deep dive is deferred in favor of getting the improved EE tracking merged first.

Current effort can be found in the branch [austinw/tracik_ee_control_bk](https://github.com/facebookresearch/fairo/tree/austinw/tracik_ee_control_bk).

## Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] Proposes a change (non-breaking change that isn't necessarily a bug)
- [ ] Refactor
- [ ] New feature (non-breaking change that adds a new functionality)
- [x] Breaking change (fix or feature that would break some existing functionality downstream)
- [ ] This is a unit test
- [ ] Documentation only change
- [ ] Datasets Release
- [ ] Models Release

## Type of requested review

- [x] I want a thorough review of the implementation.
- [ ] I want a high level review. 
- [ ] I want a deep design review.

## Before and After

If this PR introduces a change or fixes a bug, please add before and after screenshots (if this is causes a UX change) or before and after examples(this could be plain text, videos etc ) to demonstrate the change.

If this PR introduces a new feature, please also add examples or screenshots demonstrating the feature.

# Testing

Benchmark results from `benchmark_ee_tracking.py`(with Pinocchio IK because of trac-ik installation issues):
```
=== Offline (move_to_joint_positions) tracking results ===
        Pos error(mm): mean=5.43, std=4.13, max=20.64
        Ori error(deg): mean=1.30, std=0.31, max=2.01
=== Offline (move_to_ee_pose) tracking results ===
        Pos error(mm): mean=5.51, std=4.32, max=19.90
        Ori error(deg): mean=1.26, std=0.28, max=2.08
=== Online (update_desired_ee_pose) tracking results ===
        Pos error(mm): mean=4.16, std=2.78, max=14.18
        Ori error(deg): mean=1.61, std=0.28, max=2.07
```

Performance for EE pose movements were greatly improved (compared to the results in (#1307 and #1255).

# Checklist:

- [ ] I have performed manual end-to-end testing of the feature in my environment.
- [ ] I have added Docstrings and comments to the code.
- [ ] I have made changes to existing documentation where needed.
- [ ] I have added tests that show that the PR is functional.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have added relevant collaborators to review the PR before merge.
- [ ] [Polymetis only] I ran on hardware (1) all scripts in `tests/scripts`, (2) asv benchmarks.

### CODE REVIEW

File: polymetis/polymetis/python/polymetis/robot_interface.py [hunk 1]
```diff
@@ -447,6 +469,7 @@ def move_to_ee_pose(
         delta: bool = False,
         Kx: torch.Tensor = None,
         Kxd: torch.Tensor = None,
+        op_space_interp: bool = True,
         **kwargs,
     ) -> List[RobotState]:
         """Uses an operational space controller to move to a desired end-effector position (and, optionally orientation).
```
Comment (1heart) [inline polymetis/polymetis/python/polymetis/robot_interface.py:475]: Documentation here would be useful: what does op_space_interp do, as well as the fact that we're actually not doing opspace control anymore.

File: polymetis/polymetis/python/torchcontrol/planning/min_jerk.py [hunk 2]
```diff
@@ -197,6 +199,7 @@ def generate_cartesian_target_joint_min_jerk(
     """
     steps = _compute_num_steps(time_to_go, hz)
     dt = 1.0 / hz
+    home_pose = torch.zeros_like(joint_pos_start) if home_pose is None else home_pose
```
Comment (1heart) [inline polymetis/polymetis/python/torchcontrol/planning/min_jerk.py:202]: Are we sure we want the default `home_pose` to be the Franka pointed straight up, instead of e.g. the user-defined rest pose? (I understand this is what we were assuming before, so nbd if it's more annoying to fix)

File: polymetis/tests/scripts/0_joint_positions.py [hunk 3]
```diff
@@ -35,35 +35,38 @@ def test_new_joint_pos(robot, joint_pos_desired):
     print("=== RobotInterface.move_to_joint_positions ===")
     delta_joint_pos_desired = torch.Tensor([0.0, 0.0, 0.0, 0.5, 0.0, -0.5, 0.0])
     joint_pos_desired = joint_pos + delta_joint_pos_desired
+    time_to_go = 3.0
```
Comment (1heart) [inline polymetis/tests/scripts/0_joint_positions.py:38]: Not sure I understand the reasoning behind this file's changes.

File: polymetis/polymetis/torch_isolation/pinocchio_isolation/src/pinocchio_wrapper.cpp [hunk 4]
```diff
@@ -166,6 +165,13 @@ int64_t get_link_idx_from_name(State *state, const char *link_name) {
   return result;
 }
 
+char *get_link_name_from_idx(State *state, int64_t link_idx) {
+  if (link_idx >= state->model->nframes) {
```
Comment (1heart) [inline polymetis/polymetis/torch_isolation/pinocchio_isolation/src/pinocchio_wrapper.cpp:169]: Unit test this, esp. since Pinocchio's nframes is has a slightly different convention iirc.

**Review by 1heart — APPROVED** [review]
Possible improvements (maybe in a future PR):
- The Pinocchio IK doesn't respect joint limits, so we should still look for a replacement.
- Investigate a better Cartesian-space controller

Otherwise, minor nits. Thanks!

**Comment by exhaustin** [comment]
> 

Made an issue to keep track of finding better IK solutions: https://github.com/facebookresearch/fairo/issues/1336

What would we be looking for when investigating better Cartesian space controllers? Further improvement of tracking accuracy?

File: polymetis/polymetis/python/torchcontrol/planning/min_jerk.py (hunk 2 above)
Comment (exhaustin) [reply polymetis/polymetis/python/torchcontrol/planning/min_jerk.py:202]: Here I just wanted to have a sensible, platform-agnostic default within the torchcontrol module.
When calling it from RobotInterface I would always pass in the rest_pose defined within the robot model config.

File: polymetis/tests/scripts/0_joint_positions.py (hunk 3 above)
Comment (exhaustin) [reply polymetis/tests/scripts/0_joint_positions.py:38]: Before the default time_to_go in RobotInterface was set to 3 seconds because previously calling move_to_ee_pose with small movements might still result in large joint motions. Since we are actively using IK to figure out the joint motion now I shortened it to 1 second so we wouldn't need to wait 3 seconds for any tiny movement.
This test script originally does not explicitly assign time_to_go to the joint commands (thus using the default time_to_go), so I added back 3.0 seconds of time_to_go to retain the same behavior as before.

File: polymetis/polymetis/torch_isolation/pinocchio_isolation/src/pinocchio_wrapper.cpp (hunk 4 above)
Comment (exhaustin) [reply polymetis/polymetis/torch_isolation/pinocchio_isolation/src/pinocchio_wrapper.cpp:169]: Oops, this functionality is actually only needed for instantiating a trac-IK solver, but I'll add a test for it and keep it in the PR in case it comes in handy someday.
